// @contentstack/delivery-sdk/dist/cjs/src/lib/cache.js
"use strict";
// import { PersistanceStore } from '../persistance';
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleRequest = void 0;
const tslib_1 = require("tslib");
const persistance_1 = require("../persistance");
const types_1 = require("./types");
function handleRequest(cacheOptions, apiKey, defaultAdapter, resolve, reject, config) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const cacheStore = new persistance_1.PersistanceStore(cacheOptions);
        switch (cacheOptions.policy) {
            case types_1.Policy.NETWORK_ELSE_CACHE: {
                const apiResponse = yield defaultAdapter(config);
                if (apiResponse.data) {
                    cacheStore.setItem(apiKey, apiResponse.data, config.contentTypeUid, cacheOptions.maxAge);
                    return resolve(apiResponse.data);
                }
                else {
                    const cacheResponse = cacheStore.getItem(apiKey, config.contentTypeUid);
                    if (cacheResponse)
                        return resolve({
                            data: cacheResponse,
                            status: 200,
                            statusText: 'OK',
                            headers: {},
                            config: {},
                        });
                }
                return reject(apiResponse);
            }
            case types_1.Policy.CACHE_THEN_NETWORK: {
                const cacheResponse = cacheStore.getItem(apiKey, config.contentTypeUid);
                if (cacheResponse)
                    return resolve({
                        data: cacheResponse,
                        status: 200,
                        statusText: 'OK',
                        headers: {},
                        config: {},
                    });
                const apiResponse = yield defaultAdapter(config);
                if (apiResponse.data) {
                    cacheStore.setItem(apiKey, apiResponse.data, config.contentTypeUid, cacheOptions.maxAge);
                    return resolve(apiResponse.data);
                }
                else {
                    return reject(apiResponse);
                }
            }
            case types_1.Policy.CACHE_ELSE_NETWORK: {
                const cacheResponse = cacheStore.getItem(apiKey, config.contentTypeUid);
                if (cacheResponse)
                    return resolve({
                        data: cacheResponse,
                        status: 200,
                        statusText: 'OK',
                        headers: {},
                        config: {},
                    });
                else {
                    const apiResponse = yield defaultAdapter(config);
                    if (apiResponse.data) {
                        cacheStore.setItem(apiKey, apiResponse.data, config.contentTypeUid, cacheOptions.maxAge);
                        return resolve(apiResponse.data);
                    }
                    else {
                        return reject(apiResponse);
                    }
                }
            }
        }
    });
}
exports.handleRequest = handleRequest;
