// @contentstack/delivery-sdk/dist/cjs/src/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryOperator = exports.TaxonomyQueryOperation = exports.QueryOperation = exports.CanvasBy = exports.ResizeFilter = exports.OverlayRepeat = exports.OverlayAlign = exports.Orientation = exports.FitBy = exports.CropBy = exports.Format = exports.PublishType = exports.Policy = exports.Region = void 0;
var Region;
(function (Region) {
    Region["US"] = "us";
    Region["EU"] = "eu";
    Region["AZURE_NA"] = "azure-na";
    Region["AZURE_EU"] = "azure-eu";
    Region["GCP_NA"] = "gcp-na";
})(Region || (exports.Region = Region = {}));
var Policy;
(function (Policy) {
    Policy["IGNORE_CACHE"] = "IGNORE_CACHE";
    Policy["CACHE_THEN_NETWORK"] = "CACHE_THEN_NETWORK";
    Policy["CACHE_ELSE_NETWORK"] = "CACHE_ELSE_NETWORK";
    Policy["NETWORK_ELSE_CACHE"] = "NETWORK_ELSE_CACHE";
})(Policy || (exports.Policy = Policy = {}));
var PublishType;
(function (PublishType) {
    PublishType["ENTRY_PUBLISHED"] = "entry_published";
    PublishType["ENTRY_UNPUBLISHED"] = "entry_unpublished";
    PublishType["ENTRY_DELETED"] = "entry_deleted";
    PublishType["ASSET_PUBLISHED"] = "asset_published";
    PublishType["ASSET_UNPUBLISHED"] = "asset_unpublished";
    PublishType["ASSET_DELETED"] = "asset_deleted";
    PublishType["CONTENT_TYPE_DELETED"] = "content_type_deleted";
})(PublishType || (exports.PublishType = PublishType = {}));
var Format;
(function (Format) {
    Format["GIF"] = "gif";
    Format["PNG"] = "png";
    Format["JPG"] = "jpg";
    Format["PJPG"] = "pjpg";
    Format["WEBP"] = "webp";
    Format["WEBPLL"] = "webpll";
    Format["WEBPLY"] = "webply";
})(Format || (exports.Format = Format = {}));
var CropBy;
(function (CropBy) {
    CropBy["DEFAULT"] = "default";
    CropBy["ASPECTRATIO"] = "aspectRatio";
    CropBy["REGION"] = "region";
    CropBy["OFFSET"] = "offset";
})(CropBy || (exports.CropBy = CropBy = {}));
var FitBy;
(function (FitBy) {
    FitBy["BOUNDS"] = "bounds";
    FitBy["CROP"] = "crop";
})(FitBy || (exports.FitBy = FitBy = {}));
var Orientation;
(function (Orientation) {
    Orientation[Orientation["DEFAULT"] = 1] = "DEFAULT";
    Orientation[Orientation["FLIP_HORIZONTAL"] = 2] = "FLIP_HORIZONTAL";
    Orientation[Orientation["FLIP_HORIZONTAL_VERTICAL"] = 3] = "FLIP_HORIZONTAL_VERTICAL";
    Orientation[Orientation["FLIP_VERTICAL"] = 4] = "FLIP_VERTICAL";
    Orientation[Orientation["FLIP_HORIZONTAL_LEFT"] = 5] = "FLIP_HORIZONTAL_LEFT";
    Orientation[Orientation["RIGHT"] = 6] = "RIGHT";
    Orientation[Orientation["FLIP_HORIZONTAL_RIGHT"] = 7] = "FLIP_HORIZONTAL_RIGHT";
    Orientation[Orientation["LEFT"] = 8] = "LEFT";
})(Orientation || (exports.Orientation = Orientation = {}));
var OverlayAlign;
(function (OverlayAlign) {
    OverlayAlign["TOP"] = "top";
    OverlayAlign["BOTTOM"] = "bottom";
    OverlayAlign["LEFT"] = "left";
    OverlayAlign["RIGHT"] = "right";
    OverlayAlign["MIDDLE"] = "middle";
    OverlayAlign["CENTER"] = "center";
})(OverlayAlign || (exports.OverlayAlign = OverlayAlign = {}));
var OverlayRepeat;
(function (OverlayRepeat) {
    OverlayRepeat["X"] = "x";
    OverlayRepeat["Y"] = "y";
    OverlayRepeat["BOTH"] = "both";
})(OverlayRepeat || (exports.OverlayRepeat = OverlayRepeat = {}));
var ResizeFilter;
(function (ResizeFilter) {
    ResizeFilter["NEAREST"] = "nearest";
    ResizeFilter["BILINEAR"] = "bilinear";
    ResizeFilter["BICUBIC"] = "bicubic";
    ResizeFilter["LANCZOS2"] = "lanczos2";
    ResizeFilter["LANCZOS3"] = "lanczos3";
})(ResizeFilter || (exports.ResizeFilter = ResizeFilter = {}));
var CanvasBy;
(function (CanvasBy) {
    CanvasBy["DEFAULT"] = "default";
    CanvasBy["ASPECTRATIO"] = "aspectRatio";
    CanvasBy["REGION"] = "region";
    CanvasBy["OFFSET"] = "offset";
})(CanvasBy || (exports.CanvasBy = CanvasBy = {}));
var QueryOperation;
(function (QueryOperation) {
    QueryOperation["EQUALS"] = "";
    QueryOperation["NOT_EQUALS"] = "$ne";
    QueryOperation["INCLUDES"] = "$in";
    QueryOperation["EXCLUDES"] = "$nin";
    QueryOperation["IS_LESS_THAN"] = "$lt";
    QueryOperation["IS_LESS_THAN_OR_EQUAL"] = "$lte";
    QueryOperation["IS_GREATER_THAN"] = "$gt";
    QueryOperation["IS_GREATER_THAN_OR_EQUAL"] = "$gte";
    QueryOperation["EXISTS"] = "$exists";
    QueryOperation["MATCHES"] = "$regex";
})(QueryOperation || (exports.QueryOperation = QueryOperation = {}));
var TaxonomyQueryOperation;
(function (TaxonomyQueryOperation) {
    TaxonomyQueryOperation["ABOVE"] = "$above";
    TaxonomyQueryOperation["BELOW"] = "$below";
    TaxonomyQueryOperation["EQ_ABOVE"] = "$eq_above";
    TaxonomyQueryOperation["EQ_BELOW"] = "$eq_below";
})(TaxonomyQueryOperation || (exports.TaxonomyQueryOperation = TaxonomyQueryOperation = {}));
var QueryOperator;
(function (QueryOperator) {
    QueryOperator["AND"] = "$and";
    QueryOperator["OR"] = "$or";
})(QueryOperator || (exports.QueryOperator = QueryOperator = {}));
