// @contentstack/delivery-sdk/dist/esm/src/lib/stack.js
import { __awaiter } from "tslib";
import { Asset } from './asset';
import { AssetQuery } from './asset-query';
import { ContentType } from './content-type';
import { ContentTypeQuery } from './contenttype-query';
import { synchronization } from './synchronization';
import { TaxonomyQuery } from './taxonomy-query';
import { GlobalFieldQuery } from './global-field-query';
import { GlobalField } from './global-field';
export class Stack {
    constructor(client, config) {
        this._client = client;
        this.config = config;
    }
    asset(uid) {
        if (uid)
            return new Asset(this._client, uid);
        return new AssetQuery(this._client);
    }
    contentType(uid) {
        if (uid)
            return new ContentType(this._client, uid);
        return new ContentTypeQuery(this._client);
    }
    /**
     * @method Taxonomy
     * @memberOf Stack
     * @description Sets the url to /taxonomies/entries. Pass a query to fetch entries with taxonomies
     *
     * @returns {TaxonomyQuery} * @example
     * import contentstack from '@contentstack/typescript'
     *
     * const stack = contentstack.Stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
  
     * const taxonomy = stack.Taxonomy() // For taxonomy query object
     */
    taxonomy() {
        return new TaxonomyQuery(this._client);
    }
    ;
    globalField(uid) {
        if (uid)
            return new GlobalField(this._client, uid);
        return new GlobalFieldQuery(this._client);
    }
    /**
     * @method setLocale
     * @memberOf Stack
     * @description Sets the locale of the API server
     * @param {String} locale - valid locale e.g. fr-fr
     * @return {Stack}
     * @example
     * import contentstack from '@contentstack/delivery-sdk'
     *
     * const stack = contentstack.Stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
     * stack.setLocale('en-155');
     */
    setLocale(locale) {
        this.config.locale = locale;
    }
    /**
     * @method sync
     * @memberOf Stack
     * @description Syncs your Contentstack data with your app and ensures that the data is always up-to-date by providing delta updates
     * @param {object} params - params is an object that supports ‘locale’, ‘start_date’, ‘content_type_uid’, and ‘type’ queries.
     * @example
     * Stack.sync()        // For initializing sync
     * @example
     * Stack.sync({ 'locale': 'en-us'})     //For initializing sync with entries of a specific locale
     * @example
     * Stack.sync({ 'start_date': '2018-10-22'})    //For initializing sync with entries published after a specific date
     * @example
     * Stack.sync({ 'content_type_uid': 'session'})   //For initializing sync with entries of a specific content type
     * @example
     * Stack.sync({ 'type': 'entry_published'})
     * //Use the type parameter to get a specific type of content. Supports'asset_published',
     * // 'entry_published', 'asset_unpublished', 'entry_unpublished', 'asset_deleted', 'entry_deleted', 'content_type_deleted'.
     * @example
     * Stack.sync({'pagination_token': '<page_tkn>'})    // For fetching the next batch of entries using pagination token
     * @example
     * Stack.sync({'sync_token': '<sync_tkn>'})    // For performing subsequent sync after initial sync
     * @instance
     */
    sync() {
        return __awaiter(this, arguments, void 0, function* (params = {}, recursive = false) {
            return yield synchronization(this._client, params, recursive);
        });
    }
    livePreviewQuery(query) {
        if (this.config.live_preview) {
            this.config.live_preview.live_preview = query.live_preview || 'init';
            this.config.live_preview.contentTypeUid = query.contentTypeUid;
            this.config.live_preview.entryUid = query.entryUid;
        }
    }
}
