// @contentstack/management/dist/es5/app/installation/index.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _asyncToGenerator2 = require("@babel/runtime/helpers/asyncToGenerator");
var _asyncToGenerator3 = (0, _interopRequireDefault2["default"])(_asyncToGenerator2);
exports.Installation = Installation;
exports.InstallationCollection = InstallationCollection;
var _regenerator = require("@babel/runtime/regenerator");
var _regenerator2 = (0, _interopRequireDefault2["default"])(_regenerator);
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _contentstackError = require("../../core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
var _entity = require("../../entity");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function Installation(http, data) {
  var _this = this;
  var params = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  this.params = params;
  if (data.data) {
    Object.assign(this, (0, _cloneDeep2["default"])(data.data));
    if (this.installation_uid) {
      this.uid = this.installation_uid;
    }
    if (this.organization_uid) {
      this.params = {
        organization_uid: this.organization_uid
      };
    }
    this.urlPath = "/installations/".concat(this.uid);

    /**
     * @description The GET installation call is used to retrieve a specific installation of an app.
     * @memberof Installation
     * @func fetch
     * @returns {Promise<Installation>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').fetch()
     * .then((installation) => console.log(installation))
     *
     */
    this.fetch = (0, _entity.fetch)(http, 'data', this.params);

    /**
     * @description The Update installation call is used to update information of an installation.
     * @memberof Installation
     * @func update
     * @returns {Promise<Installation>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * const updateInstallation = {
     *  name: 'APP_NAME',
     *  description: 'APP_DESCRIPTION',
     *  target_type: 'stack'/'organization',
     * }
     *
     * const installation = client.organization('organization_uid').app('app_uid').installation('installation_uid')
     * installation = Object.assign(installation, updateInstallation)
     * installation.update()
     * .then((installation) => console.log(installation))
     */
    this.update = (0, _entity.update)(http, 'data', this.params);

    /**
     * @description The Uninstall installation call is used to uninstall the installation.
     * @memberof Installation
     * @func uninstall
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').uninstall()
     * .then((response) => console.log(response))
     */
    this.uninstall = (0, _entity.deleteEntity)(http, false, this.params);

    /**
     * @description To fetch organization level app installation configuration.
     * @memberof Installation
     * @func configuration
     * @param {*} param
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').configuration()
     * .then((response) => console.log(response))
     */
    this.configuration = /*#__PURE__*/(0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee() {
      var param,
        headers,
        response,
        _args = arguments;
      return _regenerator2["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            param = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
            _context.prev = 1;
            headers = {
              headers: _objectSpread({}, (0, _cloneDeep2["default"])(params)),
              params: _objectSpread({}, (0, _cloneDeep2["default"])(param))
            } || {};
            _context.next = 5;
            return http.get("".concat(_this.urlPath, "/configuration"), headers);
          case 5:
            response = _context.sent;
            if (!response.data) {
              _context.next = 10;
              break;
            }
            return _context.abrupt("return", response.data);
          case 10:
            throw (0, _contentstackError2["default"])(response);
          case 11:
            _context.next = 16;
            break;
          case 13:
            _context.prev = 13;
            _context.t0 = _context["catch"](1);
            throw (0, _contentstackError2["default"])(_context.t0);
          case 16:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 13]]);
    }));
    /**
     * @description To update organization level app installation configuration.
     * @memberof Installation
     * @func setConfiguration
     * @param {*} config Config that needs to be updated
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').setConfiguration({<configuration_details>})
     * .then((response) => console.log(response))
     */
    this.setConfiguration = /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee2(config) {
        var headers, response;
        return _regenerator2["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.prev = 0;
              headers = {
                headers: _objectSpread({}, (0, _cloneDeep2["default"])(params))
              } || {};
              _context2.next = 4;
              return http.put("".concat(_this.urlPath, "/configuration"), config, headers);
            case 4:
              response = _context2.sent;
              if (!response.data) {
                _context2.next = 9;
                break;
              }
              return _context2.abrupt("return", response.data);
            case 9:
              throw (0, _contentstackError2["default"])(response);
            case 10:
              _context2.next = 15;
              break;
            case 12:
              _context2.prev = 12;
              _context2.t0 = _context2["catch"](0);
              throw (0, _contentstackError2["default"])(_context2.t0);
            case 15:
            case "end":
              return _context2.stop();
          }
        }, _callee2, null, [[0, 12]]);
      }));
      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }();

    /**
     * @description To fetch server side organization level config required for the app.
     * @memberof Installation
     * @func getServerConfig
     * @param {*} param
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').serverConfig()
     * .then((response) => console.log(response))
     */
    this.serverConfig = /*#__PURE__*/(0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee3() {
      var param,
        headers,
        response,
        _args3 = arguments;
      return _regenerator2["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            param = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : {};
            _context3.prev = 1;
            headers = {
              headers: _objectSpread({}, (0, _cloneDeep2["default"])(params)),
              params: _objectSpread({}, (0, _cloneDeep2["default"])(param))
            } || {};
            _context3.next = 5;
            return http.get("".concat(_this.urlPath, "/server-configuration"), headers);
          case 5:
            response = _context3.sent;
            if (!response.data) {
              _context3.next = 10;
              break;
            }
            return _context3.abrupt("return", response.data);
          case 10:
            throw (0, _contentstackError2["default"])(response);
          case 11:
            _context3.next = 16;
            break;
          case 13:
            _context3.prev = 13;
            _context3.t0 = _context3["catch"](1);
            throw (0, _contentstackError2["default"])(_context3.t0);
          case 16:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[1, 13]]);
    }));
    /**
     * @description To update server side organization level config required for the app.
     * @memberof Installation
     * @func setServerConfig
     * @param {*} config Config that needs to be updated
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').setServerConfig({<configuration_details>})
     * .then((response) => console.log(response))
     */
    this.setServerConfig = /*#__PURE__*/function () {
      var _ref4 = (0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee4(config) {
        var headers, response;
        return _regenerator2["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.prev = 0;
              headers = {
                headers: _objectSpread({}, (0, _cloneDeep2["default"])(params))
              } || {};
              _context4.next = 4;
              return http.put("".concat(_this.urlPath, "/server-configuration"), config, headers);
            case 4:
              response = _context4.sent;
              if (!response.data) {
                _context4.next = 9;
                break;
              }
              return _context4.abrupt("return", response.data);
            case 9:
              throw (0, _contentstackError2["default"])(response);
            case 10:
              _context4.next = 15;
              break;
            case 12:
              _context4.prev = 12;
              _context4.t0 = _context4["catch"](0);
              throw (0, _contentstackError2["default"])(_context4.t0);
            case 15:
            case "end":
              return _context4.stop();
          }
        }, _callee4, null, [[0, 12]]);
      }));
      return function (_x2) {
        return _ref4.apply(this, arguments);
      };
    }();

    /**
     * @description To fetch installation data of an app configuration.
     * @memberof Installation
     * @func installationData
     * @returns {Promise<Response>}
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client({ authtoken: 'TOKEN'})
     *
     * client.organization('organization_uid').app('app_uid').installation('installation_uid').installationData()
     * .then((response) => console.log(response))
     */
    this.installationData = /*#__PURE__*/(0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee5() {
      var headers, response;
      return _regenerator2["default"].wrap(function _callee5$(_context5) {
        while (1) switch (_context5.prev = _context5.next) {
          case 0:
            _context5.prev = 0;
            headers = {
              headers: _objectSpread({}, (0, _cloneDeep2["default"])(params))
            } || {};
            _context5.next = 4;
            return http.get("".concat(_this.urlPath, "/installationData"), headers);
          case 4:
            response = _context5.sent;
            if (!response.data) {
              _context5.next = 9;
              break;
            }
            return _context5.abrupt("return", response.data);
          case 9:
            throw (0, _contentstackError2["default"])(response);
          case 10:
            _context5.next = 15;
            break;
          case 12:
            _context5.prev = 12;
            _context5.t0 = _context5["catch"](0);
            throw (0, _contentstackError2["default"])(_context5.t0);
          case 15:
          case "end":
            return _context5.stop();
        }
      }, _callee5, null, [[0, 12]]);
    }));
  } else {
    if (data.app_uid) {
      this.urlPath = "manifests/".concat(data.app_uid, "/installations");
      /**
       * @description The find installation call is used to retrieve all installations of your Contentstack organization.
       * @memberof Installation
       * @func findAll
       * @returns {Promise<ContentstackCollection<Installation>>}
       *
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client({ authtoken: 'TOKEN'})
       *
       * client.organization('organization_uid').app('app_uid').installation().fetchAll()
       * .then((collection) => console.log(collection))
       *
       */
      this.findAll = (0, _entity.fetchAll)(http, InstallationCollection, this.params);
    }
  }
  return this;
}
function InstallationCollection(http, data) {
  var obj = (0, _cloneDeep2["default"])(data.data) || [];
  return obj.map(function (installationData) {
    return new Installation(http, {
      data: installationData
    });
  });
}
