// @contentstack/management/dist/es5/stack/asset/index.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _asyncToGenerator2 = require("@babel/runtime/helpers/asyncToGenerator");
var _asyncToGenerator3 = (0, _interopRequireDefault2["default"])(_asyncToGenerator2);
exports.Asset = Asset;
exports.AssetCollection = AssetCollection;
exports.createFormData = createFormData;
var _regenerator = require("@babel/runtime/regenerator");
var _regenerator2 = (0, _interopRequireDefault2["default"])(_regenerator);
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _entity = require("../../entity");
var _folders = require("./folders");
var _contentstackError = require("../../core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
var _formData = require("form-data");
var _formData2 = (0, _interopRequireDefault2["default"])(_formData);
var _fs = require("fs");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty3["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
/**
 * Assets refer to all the media files (images, videos, PDFs, audio files, and so on) uploaded in your Contentstack repository for future use.
 * These files can be attached and used in multiple entries. Read more about <a href='https://www.contentstack.com/docs/guide/content-management'>Assets</a>.
 * @namespace Asset
 */

function Asset(http) {
  var _this = this;
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  this.stackHeaders = data.stackHeaders;
  this.urlPath = "/assets";
  if (data.asset) {
    Object.assign(this, (0, _cloneDeep2["default"])(data.asset));
    this.urlPath = "/assets/".concat(this.uid);

    /**
     * @description The Update Asset call lets you update the name and description of an existing Asset.
     * @memberof Asset
     * @func update
     * @returns {Promise<Asset.Asset>} Promise for Asset instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').fetch()
     * .then((asset) => {
     *  asset.title = 'My New asset'
     *  asset.description = 'Asset description'
     *  return asset.update()
     * })
     * .then((asset) => console.log(asset))
     *
     */
    this.update = (0, _entity.update)(http, 'asset');

    /**
     * @description The Delete asset call will delete an existing asset from the stack.
     * @memberof Asset
     * @func delete
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').delete()
     * .then((response) => console.log(response.notice))
     */
    this["delete"] = (0, _entity.deleteEntity)(http);

    /**
     * @description The fetch an asset call returns comprehensive information about a specific version of an asset of a stack.
     * @memberof Asset
     * @func fetch
     * @returns {Promise<Asset.Asset>} Promise for Asset instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').fetch()
     * .then((asset) => console.log(asset))
     *
     */
    this.fetch = (0, _entity.fetch)(http, 'asset');

    /**
     * @description The Replace asset call will replace an existing asset with another file on the stack.
     * @memberof Asset
     * @func replace
     * @returns {Promise<Asset.Asset>} Promise for Asset instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const asset = {
     *  upload: 'path/to/file.png',
     * }
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').replace(asset)
     * .then((asset) => console.log(asset))
     *
     */
    this.replace = /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee(data, params) {
        var response;
        return _regenerator2["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 3;
              return (0, _entity.upload)({
                http: http,
                urlPath: this.urlPath,
                stackHeaders: this.stackHeaders,
                formData: createFormData(data),
                params: params,
                method: 'PUT'
              });
            case 3:
              response = _context.sent;
              if (!response.data) {
                _context.next = 8;
                break;
              }
              return _context.abrupt("return", new this.constructor(http, (0, _entity.parseData)(response, this.stackHeaders)));
            case 8:
              throw (0, _contentstackError2["default"])(response);
            case 9:
              _context.next = 14;
              break;
            case 11:
              _context.prev = 11;
              _context.t0 = _context["catch"](0);
              throw (0, _contentstackError2["default"])(_context.t0);
            case 14:
            case "end":
              return _context.stop();
          }
        }, _callee, this, [[0, 11]]);
      }));
      return function (_x, _x2) {
        return _ref.apply(this, arguments);
      };
    }();
    /**
     * @description The Publish an asset call is used to publish a specific version of an asset on the desired environment either immediately or at a later date/time.
     * @memberof Asset
     * @func publish
     * @returns {Promise<Object>} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const asset = {
     *  "locales": [
     *              "en-us"
     *              ],
     *   "environments": [
     *                "development"
     *               ]
     * }
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').publish({ publishDetails: asset, version: 1, scheduledAt: "2019-02-08T18:30:00.000Z"})
     * .then((response) => console.log(response.notice))
     *
     */
    this.publish = (0, _entity.publish)(http, 'asset');

    /**
     * @description The Replace asset call will replace an existing asset with another file on the stack.
     * @memberof Asset
     * @func unpublish
     * @returns {Promise<Object>} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const asset = {
     *  "locales": [
     *              "en-us"
     *              ],
     *   "environments": [
     *                "development"
     *               ]
     * }
     *
     * client.stack({ api_key: 'api_key'}).asset('uid').unpublish({ publishDetails: asset, version: 1, scheduledAt: "2019-02-08T18:30:00.000Z"})
     * .then((response) => console.log(response.notice))
     *
     */
    this.unpublish = (0, _entity.unpublish)(http, 'asset');
  } else {
    /**
     *  @description The Folder allows to fetch and create folders in assets.
     * @memberof Asset
     * @func folder
     * @returns {Promise<Folder.Folder>} Promise for Entry instance
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const asset = {name: 'My New contentType'}
     * client.stack({ api_key: 'api_key'}).asset('uid').folder().create({ asset })
     * .then((folder) => console.log(folder))
    */
    this.folder = function () {
      var folderUid = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      var data = {
        stackHeaders: _this.stackHeaders
      };
      if (folderUid) {
        data.asset = {
          uid: folderUid
        };
      }
      return new _folders.Folder(http, data);
    };

    /**
     * @description The Create an asset call creates a new asset.
     * @memberof Asset
     * @func create
     * @returns {Promise<Asset.Asset>} Promise for Asset instance
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const asset = {
     *  upload: 'path/to/file.png',
     *  title: 'Title',
     *  description: 'Desc'
     * }
     *
     * client.stack({ api_key: 'api_key'}).asset().create(asset)
     * .then((asset) => console.log(asset))
     */
    this.create = /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee2(data, params) {
        var response;
        return _regenerator2["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.prev = 0;
              _context2.next = 3;
              return (0, _entity.upload)({
                http: http,
                urlPath: this.urlPath,
                stackHeaders: this.stackHeaders,
                formData: createFormData(data),
                params: params
              });
            case 3:
              response = _context2.sent;
              if (!response.data) {
                _context2.next = 8;
                break;
              }
              return _context2.abrupt("return", new this.constructor(http, (0, _entity.parseData)(response, this.stackHeaders)));
            case 8:
              throw (0, _contentstackError2["default"])(response);
            case 9:
              _context2.next = 14;
              break;
            case 11:
              _context2.prev = 11;
              _context2.t0 = _context2["catch"](0);
              throw (0, _contentstackError2["default"])(_context2.t0);
            case 14:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this, [[0, 11]]);
      }));
      return function (_x3, _x4) {
        return _ref2.apply(this, arguments);
      };
    }();
    /**
     * @description The Query on Asset will allow to fetch details of all or specific Asset.
     * @memberof Asset
     * @param {Object} params - URI parameters
     * @prop {Object} params.query - Queries that you can use to fetch filtered results.
     * @func query
     * @returns {Array<Asset>} Array of Asset.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).asset().query({ query: { filename: 'Asset Name' } }).find()
     * .then((asset) => console.log(asset))
     */
    this.query = (0, _entity.query)({
      http: http,
      wrapperCollection: AssetCollection
    });
  }
  /**
   * @description The Download function will get downloadable file in specified format.
   * @memberof Asset
   * @func download
   * @returns {Array<AssetResponse>} Array of Asset.
   * @param {*} param.url The url for the asset to download
   * @param {*} param.responseType Optional parameter to specify the response type.
   * @example
   *
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.stack({ api_key: 'api_key'}).asset('uid').fetch()
   * .then((asset) => asset.download({responseType: 'blob'}))
   * .then((response) => // Write response data to destination file. )
   * @example
   *
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.stack({ api_key: 'api_key'}).asset().download({url: 'asset_url_to_download', responseType: 'blob'})
   * .then((response) => // Write response data to destination file. )
   */
  this.download = /*#__PURE__*/function () {
    var _ref4 = (0, _asyncToGenerator3["default"])( /*#__PURE__*/_regenerator2["default"].mark(function _callee3(_ref3) {
      var url, responseType, params, headers, requestUrl;
      return _regenerator2["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            url = _ref3.url, responseType = _ref3.responseType, params = _ref3.params;
            _context3.prev = 1;
            headers = {
              headers: _objectSpread(_objectSpread({}, params), (0, _cloneDeep2["default"])(this.stackHeaders)),
              responseType: responseType
            } || {
              responseType: responseType
            };
            requestUrl = url || this.url;
            if (!(!requestUrl || requestUrl === undefined)) {
              _context3.next = 6;
              break;
            }
            throw new Error('Asset URL can not be empty');
          case 6:
            return _context3.abrupt("return", http.get(requestUrl, headers));
          case 9:
            _context3.prev = 9;
            _context3.t0 = _context3["catch"](1);
            throw (0, _contentstackError2["default"])(_context3.t0);
          case 12:
          case "end":
            return _context3.stop();
        }
      }, _callee3, this, [[1, 9]]);
    }));
    return function (_x5) {
      return _ref4.apply(this, arguments);
    };
  }();
  return this;
}
function AssetCollection(http, data) {
  var obj = (0, _cloneDeep2["default"])(data.assets) || [];
  var assetCollection = obj.map(function (userdata) {
    return new Asset(http, {
      asset: userdata,
      stackHeaders: data.stackHeaders
    });
  });
  return assetCollection;
}
function createFormData(data) {
  return function () {
    var formData = new _formData2["default"]();
    if (typeof data['parent_uid'] === 'string') {
      formData.append('asset[parent_uid]', data['parent_uid']);
    }
    if (typeof data.description === 'string') {
      formData.append('asset[description]', data.description);
    }
    if (data.tags instanceof Array) {
      formData.append('asset[tags]', data.tags.join(','));
    } else if (typeof data.tags === 'string') {
      formData.append('asset[tags]', data.tags);
    }
    if (typeof data.title === 'string') {
      formData.append('asset[title]', data.title);
    }
    var uploadStream = (0, _fs.createReadStream)(data.upload);
    if (typeof data.content_type === 'string') {
      formData.append('asset[upload]', uploadStream, {
        contentType: data.content_type
      });
    } else {
      formData.append('asset[upload]', uploadStream);
    }
    return formData;
  };
}
